### PR TITLE
[#3607] Refactor validation plugin, add BRK validator

### LINKED ISSUE #3607: Als gemeente wil ik user-input valideren dmv data uit het Basisregistratie Kadaster (BRK)
Als formulierontwikkelaar wil ik een koppeling maken met het BRK, om bepaalde user-input te valideren met de data die in het BRK opgeslagen staat.

Ref: DH 609

**Voorbeeld**

Formulier: Heruitgifte erfpacht aanvragen

- User logt in met DigiD
- Controle BRK: is de naam van de aanvrager (volgens inlog DigiD/BRP) van de aanvrager gelijk aan de geregistreerde zakelijk gerechtigde voor het adres waarvoor de aanvraag geldt in de BRK?

Indien correct: user kan doorgaan naar volgende stap

Indien niet correct, toon foutmelding:

Uit controle van uw gegevens blijkt dat u deze aanvraag niet kunt doen. Voor meer informatie neem contact op [...]

- [x] https://github.com/open-formulieren/types/issues/29 / https://github.com/open-formulieren/types/pull/30
- [x] https://github.com/open-formulieren/formio-builder/issues/65 / https://github.com/open-formulieren/formio-builder/pull/68
- [x] https://github.com/open-formulieren/open-forms/pull/3646
- [x] https://github.com/open-formulieren/open-forms-sdk/pull/623
- [x] Implement formatter for custom component type (backend) for PDF, email, CLI...)
- [x] Implement `address` component in SDK
- [x] Implement BRK service integration
- [x] Implement backend validator to check provided address with authenticated user, in the BRK

### PR DESCRIPTION
**Pull request by Viicos**
Partly closes #3607

### CODE REVIEW

**Comment by sergei-maertens** [comment]
Blocked by adding the formio-builder v0.11.0 in a separate PR

File: src/openforms/validations/registry.py
Comment (sergei-maertens) [reply src/openforms/validations/registry.py:98]: This should also be checked at the type level, so perhaps `ValidatorType` should be defined as a `Protocol` instead of a plain `Callable`.

File: src/openforms/validations/registry.py
Comment (sergei-maertens) [inline src/openforms/validations/registry.py:148]: The response data shape is different for both validation stages.

HTTP 400 response now has validation error information, but the SDK is not built to handle that.
HTTP 200 response contains a top level key which designates whether the result is valid or not.

The SDK needs to be able to handle those HTTP 400s and place the validation errors with the respective (nested) fields, which is going to be... a challenge, I think. :grimacing:

File: src/openforms/validations/tests/test_registry.py
Comment (sergei-maertens) [inline src/openforms/validations/tests/test_registry.py:62]: oh this only concerns tests to set the attribute? If we don't have any actual function-based validator implementations, just drop support for it.

**Review by sergei-maertens — CHANGES_REQUESTED** [review]

File: src/openforms/contrib/brk/validators.py
Comment (joeribekker) [reply src/openforms/contrib/brk/validators.py:64]: Let's keep a generic error message so this component is useful in more than 1 case.

If this is unacceptable, we'll hear about it no doubt.

File: src/openforms/contrib/brk/tests/test_validators.py
Comment (Viicos) [reply src/openforms/contrib/brk/tests/test_validators.py:20]: True, + having a setUp defined in the mixins is pretty useless currently with only one test method..

File: src/openforms/contrib/brk/validators.py
Comment (Viicos) [reply src/openforms/contrib/brk/validators.py:84]: Test added

File: src/openforms/contrib/brk/validators.py
Comment (Viicos) [reply src/openforms/contrib/brk/validators.py:91]: Made some changes but not sure I follow you here. We can discuss this on Tuesday

File: src/openforms/contrib/brk/validators.py
Comment (Viicos) [reply src/openforms/contrib/brk/validators.py:94]: Not really testable against the test API instance (as said in the comment, it shouldn't happen as the postcode and house number is supposed to uniquely identify one property).

If you want it to be covered, I'll have to manually edit cassette files

File: src/openforms/contrib/brk/validators.py
Comment (Viicos) [reply src/openforms/contrib/brk/validators.py:113]: > Auth attribute hashing only takes place for non-active submissions

Wasn't aware of this

File: src/openforms/contrib/brk/validators.py
Comment (Viicos) [reply src/openforms/contrib/brk/validators.py:118]: > I don't see any benefit of including `KeyError` and `TypeError` here.

Looking at it again it is a leftover from a previous solution I had without `glom`. The code is now safe, except here:
https://github.com/open-formulieren/open-forms/blob/ae52b85449d017785eac8c314734dd0b8d761758/src/openforms/contrib/brk/validators.py#L103-L106

but if we have exceptions with the above code, probably better to leave them unhandled as this would indicate something wrong from the BRK API.

Comment (Viicos) [reply src/openforms/contrib/brk/validators.py:118]: `try..catch` removed and ctx manager added

File: src/openforms/contrib/kvk/tests/test_validators.py
Comment (Viicos) [reply src/openforms/contrib/kvk/tests/test_validators.py:68]: Yeah was annoyed by this one haha

File: src/openforms/validations/registry.py
Comment (Viicos) [reply src/openforms/validations/registry.py:17]: (will be refactored anyway in the other PR)

File: src/openforms/validations/registry.py
Comment (Viicos) [reply src/openforms/validations/registry.py:65]: Fixed it but still not happy with it. The thing is the frontend sends this kind of data:

```json
{
    "postCode": "...",
    "houseNumber": "...",
    "houseLetter": "",
    "houseNumberAddition": ""
}
```

That is it keeps empty strings for fields that were not filled by the user, thus failing with the serializer (I think because of the regex patterns) if not filtered

File: src/openforms/contrib/brk/validators.py
Comment (sergei-maertens) [reply src/openforms/contrib/brk/validators.py:94]: if it shouldn't happen, then there also shouldn't be graceful handling IMO.

In this case, a simple `assert len(kadastraals) <= 1` is sufficient. If it crashes, it crashes and then we get a proper traceback in Sentry and need to fix our implementation because in that case we made a programming mistake.

File: src/openforms/contrib/brk/validators.py
Comment (sergei-maertens) [reply src/openforms/contrib/brk/validators.py:118]: I also tend to do the low-level processing on the API client classes, and let them return dataclasses suited to the data we're using/modelling. That also makes it type-checker friendlier. In a validator I don't really want to deal with the peculiarities of JSON+hal, that's the job of the API client who can translate it into domain objects.

I think the Haal Centraal/NP family members perform similar normalizations where it makes even more sense because there are different implementations that should also provide the same interface.

File: src/openforms/validations/registry.py
Comment (sergei-maertens) [reply src/openforms/validations/registry.py:65]: well then your serializer needs to be updated to allow empty values :)

probably you have `required=True` set because that's the default and it's tripping up things.

File: src/openforms/validations/registry.py
Comment (Viicos) [reply src/openforms/validations/registry.py:98]: > shouldn't that be `TypeError`?

I think I went with `ValueError` as even if the type of `call` is correct (i.e. a callable), it does not comply with what we expect. An alternative example:
```python
def func(a: int):
    if not isinstance(a, int):
        raise TypeError()
    if a > 0:
        raise ValueError("a should be <= 0")
```

I think both are valid.

I'll go with an assert statement, but the rest is refactored in the other PR anyway.

File: src/openforms/validations/registry.py
Comment (Viicos) [reply src/openforms/validations/registry.py:143]: See https://github.com/open-formulieren/open-forms/pull/3646/files#r1438332325, we can discuss this on Tuesday

File: src/openforms/validations/tests/test_registry.py
Comment (Viicos) [reply src/openforms/validations/tests/test_registry.py:62]: I think the next refactor PR handles this

File: src/openforms/validations/registry.py
Comment (sergei-maertens) [reply src/openforms/validations/registry.py]: yeah but the type requirement is wrong here :) that's why I suggest the `Protocol` approach since you can add additional constraints.

File: src/openforms/validations/registry.py
Comment (Viicos) [reply src/openforms/validations/registry.py]: `required` is set correctly to `False`, what I need is something like _if provided but empty string, handle the field as being not provided, otherwise apply normal validation against the regex pattern_

File: src/openforms/validations/api/views.py
Comment (Viicos) [reply src/openforms/validations/api/views.py:70]: This was mistakenly set as `CharField`, but shouldn't `JSONField` do the trick? (as discussed a while ago on Slack)

File: src/openforms/validations/registry.py
Comment (sergei-maertens) [reply src/openforms/validations/registry.py]:
```py
>>> from rest_framework import serializers
>>> class Test(serializers.Serializer):
...     optional = serializers.RegexField(allow_blank=True, regex=r"\d+")
...     required = serializers.RegexField(allow_blank=False, regex=r"\d+")
...     
... 
>>> ser = Test(data={"optional": "", "required": "123"})
>>> ser.is_valid()
True
```

File: src/openforms/validations/api/views.py
Comment (sergei-maertens) [reply src/openforms/validations/api/views.py]: I don't think it will, I think it'll be turned into

```yaml
type: object
additionalProperties: true
```
because of drf-spectacular's handling:

```py
        if isinstance(field, serializers.JSONField):
            return append_meta(build_basic_type(OpenApiTypes.OBJECT), meta)
```
in `drf_spectacular.openapi.AutoSchema._map_serializer_field`

Comment (Viicos) [reply src/openforms/validations/api/views.py]: Ah true, but isn't spectacular wrong in this case?

Comment (sergei-maertens) [reply src/openforms/validations/api/views.py]: yes :-) feel free to make an issue

It may also depend on the DB driver. Perhaps other databases only support objects for their JSON type? Might be good to raise an issue with drf-spectacular!

File: src/openforms/formio/formatters/custom.py
Comment (sergei-maertens) [inline src/openforms/formio/formatters/custom.py:31]: I think making `value` a `TypedDict` makes sense here to document the expected types, since if you put in any dict of str -> str mapping, you will get `KeyError` on the `.format` call:

```py
>>> "{foo}".format()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'foo'
```

File: src/openforms/validations/tests/test_api.py
Comment (sergei-maertens) [inline src/openforms/validations/tests/test_api.py:173]: please use the `SubmissionsMixin` and call `_add_submission_to_session`

File: src/openforms/validations/tests/test_registry.py
Comment (sergei-maertens) [inline src/openforms/validations/tests/test_registry.py:124]: should be submission instance, no?

**Review by sergei-maertens — CHANGES_REQUESTED** [review]

File: src/openforms/validations/tests/test_registry.py
Comment (Viicos) [reply src/openforms/validations/tests/test_registry.py:124]: a leftover from the first pass

File: src/openforms/contrib/brk/validators.py
Comment (Viicos) [reply src/openforms/contrib/brk/validators.py:52]: I added a test specific to this serializer, as the current brk tests makes direct use of the plugin's `__call__` method.

It could be added to the `openforms.validations.tests` module, but I feel it doesn't belong there as this test module is "generic", it shouldn't test specific plugin functionality.

Comment (sergei-maertens) [reply src/openforms/contrib/brk/validators.py:52]: I agree with the sentiment/concerns - however a test or test case in `openforms.validation.tests` decorated with `@tag("gh-3607")` makes the origin of the test clear. It proves that the feature as a whole works as intended (so it's an integration test).

I'm all for proper abstractions and loose coupling, paired with their respective unit tests, but all those bits combined can still lead to a non-working total solution (classic image of sink + pipes being misalinged ;) ).

Probably the middleground here is inside the `contrib.brk` package to add an API/integration test that calls the validate endpoint specifically with this validator. It also fits our architecture/module organization, as the `forms` and `submissions` apps are considered "core" and may be used by any other Open Forms module.

File: src/openforms/contrib/brk/tests/test_validators.py
Comment (sergei-maertens) [inline src/openforms/contrib/brk/tests/test_validators.py:30]: See the other remark, I'd rather see a new test module created: `test_integration.py` where you set up a submission and call the validation endpoint with this validator. The implementation details on the serializer are irrelevant, what you care about is that only normalized data is sent to the BRK API & produces useful results/outcomes.

**Review by sergei-maertens — APPROVED** [review]
Nice, now as a final step before this can get merged, a cleanup of the commit history is in order I think :grimacing:

**Pull request merged**
